diff --git a/sys/hps_io.sv b/sys/hps_io.sv
index 92cdf77..c7ece78 100644
--- a/sys/hps_io.sv
+++ b/sys/hps_io.sv
@@ -47,7 +47,7 @@ module hps_io #(parameter CONF_STR, CONF_STR_BRAM=0, PS2DIV=0, WIDE=0, VDNUM=1,
 	output reg [15:0] joystick_l_analog_3,
 	output reg [15:0] joystick_l_analog_4,
 	output reg [15:0] joystick_l_analog_5,
-
+	input      [15:0] joy_raw,
 	output reg [15:0] joystick_r_analog_0,
 	output reg [15:0] joystick_r_analog_1,
 	output reg [15:0] joystick_r_analog_2,
@@ -349,6 +349,8 @@ always@(posedge clk_sys) begin : uio_block
 		end else begin
 
 			casex(cmd)
+				// Reading user_io raw joy
+				'h0f: io_dout <= joy_raw;
 				// buttons and switches
 				'h01: cfg <= io_din;
 				'h02: if(byte_cnt==1) joystick_0[15:0] <= io_din; else joystick_0[31:16] <= io_din;
diff --git a/sys/sys.tcl b/sys/sys.tcl
index 93b6247..1fe1b07 100644
--- a/sys/sys.tcl
+++ b/sys/sys.tcl
@@ -28,6 +28,7 @@ set_instance_assignment -name CURRENT_STRENGTH_NEW "MAXIMUM CURRENT" -to IO_S*
 #============================================================
 # USER PORT
 #============================================================
+set_location_assignment PIN_AE15 -to USER_IO[7]
 set_location_assignment PIN_AF17 -to USER_IO[6]
 set_location_assignment PIN_AF15 -to USER_IO[5]
 set_location_assignment PIN_AG16 -to USER_IO[4]
@@ -100,7 +101,7 @@ set_instance_assignment -name ALLOW_SYNCH_CTRL_USAGE OFF -to *|SDRAM_*
 #============================================================
 # SPI SD
 #============================================================
-set_location_assignment PIN_AE15 -to SD_SPI_CS
+#set_location_assignment PIN_AE15 -to SD_SPI_CS
 set_location_assignment PIN_AH8  -to SD_SPI_MISO
 set_location_assignment PIN_AG8  -to SD_SPI_CLK
 set_location_assignment PIN_U13  -to SD_SPI_MOSI
diff --git a/sys/sys_top.v b/sys/sys_top.v
index 2b726ef..dd37179 100644
--- a/sys/sys_top.v
+++ b/sys/sys_top.v
@@ -96,7 +96,7 @@ module sys_top
 `endif
 
 	////////// I/O ALT /////////
-	output        SD_SPI_CS,
+	//output        SD_SPI_CS,
 	input         SD_SPI_MISO,
 	output        SD_SPI_CLK,
 	output        SD_SPI_MOSI,
@@ -121,7 +121,7 @@ module sys_top
 	output  [7:0] LED,
 
 	///////// USER IO ///////////
-	inout   [6:0] USER_IO
+	inout   [7:0] USER_IO
 );
 
 //////////////////////  Secondary SD  ///////////////////////////////////
@@ -221,7 +221,7 @@ always @(posedge FPGA_CLK2_50) begin
 		if(&deb_user) btn_user <= 1;
 		if(!deb_user) btn_user <= 0;
 
-		deb_osd <= {deb_osd[6:0], btn_o | ~KEY[0]};
+		deb_osd <= {deb_osd[6:0], btn_o | user_osd | ~KEY[0]};
 		if(&deb_osd) btn_osd <= 1;
 		if(!deb_osd) btn_osd <= 0;
 	end
@@ -1632,21 +1632,23 @@ audio_out audio_out
 
 ////////////////  User I/O (USB 3.0 connector) /////////////////////////
 
-assign USER_IO[0] =                       !user_out[0]  ? 1'b0 : 1'bZ;
-assign USER_IO[1] =                       !user_out[1]  ? 1'b0 : 1'bZ;
+assign USER_IO[0] = |user_mode   ? user_out[0] : !user_out[0]  ? 1'b0 : 1'bZ;
+assign USER_IO[1] = user_mode[0] ? user_out[1] : !user_out[1]  ? 1'b0 : 1'bZ;
 assign USER_IO[2] = !(SW[1] ? HDMI_I2S   : user_out[2]) ? 1'b0 : 1'bZ;
 assign USER_IO[3] =                       !user_out[3]  ? 1'b0 : 1'bZ;
-assign USER_IO[4] = !(SW[1] ? HDMI_SCLK  : user_out[4]) ? 1'b0 : 1'bZ;
+assign USER_IO[4] = user_mode[1] ? user_out[4] : !(SW[1] ? HDMI_SCLK  : user_out[4]) ? 1'b0 : 1'bZ;
 assign USER_IO[5] = !(SW[1] ? HDMI_LRCLK : user_out[5]) ? 1'b0 : 1'bZ;
 assign USER_IO[6] =                       !user_out[6]  ? 1'b0 : 1'bZ;
+assign USER_IO[7] =                       !user_out[7]  ? 1'b0 : 1'bZ;
 
-assign user_in[0] =         USER_IO[0];
-assign user_in[1] =         USER_IO[1];
+assign user_in[0] = |user_mode   ? 1'b0 : USER_IO[0];
+assign user_in[1] = user_mode[0] ? 1'b0 : USER_IO[1];
 assign user_in[2] = SW[1] | USER_IO[2];
 assign user_in[3] =         USER_IO[3];
-assign user_in[4] = SW[1] | USER_IO[4];
+assign user_in[4] = user_mode[1] ? 1'b0 : SW[1] | USER_IO[4];
 assign user_in[5] = SW[1] | USER_IO[5];
 assign user_in[6] =         USER_IO[6];
+assign user_in[7] =         USER_IO[7];
 
 
 ///////////////////  User module connection ////////////////////////////
@@ -1681,7 +1683,9 @@ wire  [1:0] btn;
 sync_fix sync_v(clk_vid, vs_emu, vs_fix);
 sync_fix sync_h(clk_vid, hs_emu, hs_fix);
 
-wire  [6:0] user_out, user_in;
+wire  [7:0] user_out, user_in;
+wire  [1:0] user_mode;
+wire        user_osd;
 
 assign clk_ihdmi= clk_vid;
 assign ce_hpix  = vga_ce_sl;
@@ -1849,6 +1853,8 @@ emu emu
 	.UART_DTR(uart_dsr),
 	.UART_DSR(uart_dtr),
 
+	.USER_OSD(user_osd),
+	.USER_MODE(user_mode),
 	.USER_OUT(user_out),
 	.USER_IN(user_in)
 );
